File: app/src/main/java/ru/merkulyevsasha/news/data/db/NewsDbRepositoryImpl.java
package ru.merkulyevsasha.news.data.db;

import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;
import ru.merkulyevsasha.news.data.ListUtils;
import ru.merkulyevsasha.news.pojos.Article;

public class NewsDbRepositoryImpl implements NewsDbRepository {

    private final NewsDbRoom room;

    private final ArticleMapper articleMapper;
    private final ArticleEntityMapper articleEntityMapper;

    @Inject
    public NewsDbRepositoryImpl(NewsDbRoom room) {
        this.room = room;
        articleMapper = new ArticleMapper();
        articleEntityMapper = new ArticleEntityMapper();
    }

    @Override
    public void addListNews(List<Article> items) {
        ArticleDao dao = room.getArticleDao();
        dao.addListNews(ListUtils.convertToArray(items, articleMapper, ArticleEntity.class));
    }

    @Override
    public void delete(final int navId) {
        ArticleDao dao = room.getArticleDao();
        dao.delete(navId);
    }

    @Override
    public void deleteAll() {
        ArticleDao dao = room.getArticleDao();
        dao.deleteAll();
    }

    @Override
    public Single<List<Article>> selectAll() {
        ArticleDao dao = room.getArticleDao();
        return dao.selectAll().flattenAsFlowable(t -> t).map(articleEntityMapper::map).toList();
    }

    @Override
    public Single<List<Article>> selectNavId(int navId) {
        ArticleDao dao = room.getArticleDao();
        return dao.selectNavId(navId).flattenAsFlowable(t -> t).map(articleEntityMapper::map).toList();
    }

    @Override
    public Single<List<Article>> search(String searchTtext) {
        ArticleDao dao = room.getArticleDao();
        return dao.search(searchTtext).flattenAsFlowable(t -> t).map(articleEntityMapper::map).toList();
    }

    @Override
    public Date getLastPubDate() {
        ArticleDao dao = room.getArticleDao();
        ArticleEntity article = dao.getLastArticle();
        return article == null? null : article.getPubDate();
    }

}

File: app/src/main/java/ru/merkulyevsasha/news/data/NewsDbRepositoryImpl.kt
package ru.merkulyevsasha.news.data

import io.reactivex.Single
import ru.merkulyevsasha.news.data.db.NewsDbRoom
import ru.merkulyevsasha.news.data.db.entities.ArticleEntity
import ru.merkulyevsasha.news.data.db.mappers.ArticleEntityMapper
import ru.merkulyevsasha.news.data.db.mappers.ArticleMapper
import ru.merkulyevsasha.news.models.Article
import java.util.*
import javax.inject.Inject

class NewsDbRepositoryImpl @Inject
constructor(private val room: NewsDbRoom) : NewsDbRepository {

    private val articleMapper: ArticleMapper
    private val articleEntityMapper: ArticleEntityMapper

    override val lastPubDate: Date?
        get() {
            val dao = room.articleDao
            val article = dao.getLastArticle()
            return article.pubDate
        }

    init {
        articleMapper = ArticleMapper()
        articleEntityMapper = ArticleEntityMapper()
    }

    override fun addListNews(items: List<Article>) {
        val dao = room.articleDao
        dao.addListNews(*MappersUtils.convertToArray(items, articleMapper, ArticleEntity::class.java))
    }

    override fun delete(navId: Int) {
        val dao = room.articleDao
        dao.delete(navId)
    }

    override fun deleteAll() {
        val dao = room.articleDao
        dao.deleteAll()
    }

    override fun selectAll(): Single<List<Article>> {
        val dao = room.articleDao
        return dao.selectAll().flattenAsFlowable { t -> t }.map<Article> { articleEntityMapper.map(it) }.toList()
    }

    override fun selectNavId(navId: Int): Single<List<Article>> {
        val dao = room.articleDao
        return dao.selectNavId(navId).flattenAsFlowable { t -> t }.map<Article> { articleEntityMapper.map(it) }.toList()
    }

    override fun search(searchTtext: String): Single<List<Article>> {
        val dao = room.articleDao
        return dao.search(searchTtext).flattenAsFlowable { t -> t }.map<Article> { articleEntityMapper.map(it) }.toList()
    }

}
